{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "43a8c65b",
   "metadata": {},
   "source": [
    "# Installing Dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "800a489c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np # zero matrices, arrays, .max(), linspace\n",
    "from numpy.random import randint # random integers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b65b939",
   "metadata": {},
   "source": [
    "# Cleaning Up Matrix For Printing\n",
    "\n",
    "The first function add_bases() employs three parameters (x = first sequence, y = second sequence, m = current matrix). This function adds genetic sequences on the side of the row and column to allow for easier formatting.\n",
    "\n",
    "Before implementing add_bases():\n",
    "![unfilled_version.png](Unfilled_version.png)\n",
    "\n",
    "After implementing add_bases()\n",
    "![filled_version.png](filled_version.png)\n",
    "\n",
    "The second function print_matrix() has one parameter (matrix = current matrix). It employs methods to allow for better spacing and line formatting for the array itself (evident in second picture)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "766509b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_bases (x, y, m):\n",
    "    matrix = m.tolist()\n",
    "    matrix.insert(0, [\" \",  \" \"]+list(y)) # Adding the genetic sequence to the top row\n",
    "    for i in range(2, len(x) + 2): # Adding the genetic sequence to the leftmost column\n",
    "        matrix[i] = [list(x)[i-2]] + matrix[i]\n",
    "    matrix[1] = [\" \"] + matrix[1]\n",
    "    return matrix\n",
    "    \n",
    "def print_matrix(matrix):\n",
    "    s = [[str(e) for e in row] for row in matrix]\n",
    "    lens = [max(map(len, col)) for col in zip(*s)]\n",
    "    fmt = '\\t'.join('{{:{}}}'.format(x) for x in lens)\n",
    "    table = [fmt.format(*row) for row in s]\n",
    "    print('\\n'.join(table))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ba11b56",
   "metadata": {},
   "source": [
    "# Needleman-Wunsch Algorithm\n",
    "\n",
    "The Needleman-Wunsch algorithm is run through the nw() function with four parameters (x: the first sequence, y: the second sequence, mismatch: the penalty for mismatching two sequences, gap: the penalty for an insertion/deletion).\n",
    "\n",
    "Global alignment refers to trying to get every element inside two particular genetic sequences to align with one another; this is often used to calculate evolutionary processes that were acted to preserve biological function through changing a specific codon/nucleotide (not by random chance). The main purpose of the Needleman-Wunsch algorithm is to discover how one sequence can relate to another with mutations (change in nitrogenous bases), additions, or deletions of nitrogenous base pairs.\n",
    "\n",
    "Global alignment:\n",
    "![global_alignment.png](global_alignment.png)\n",
    "\n",
    "For example, the algorithm would identify the sequences “CGC” and “ACGC” to be evolutionarily related, with a simple addition of adenine to get “CGC” from “ACGC”. Of course, for more complex sequences, a more optimal algorithm is necessary.\n",
    "\n",
    "The main goal of the N-W algorithm is to minimize the number of events to explain the differences between two nucleotide sequences, by running this algorithm from exponential to quadratic runtime using dynamic programming. It would recursively fill up a table, discover optimal scores, then trace back the optimal score to then find the optimal solution to the code. This is done through a base matrix (which simply runs forward the dynamic programming algorithm) and then a subsequent pointer matrix (which tries to run back into the code to figure out the most optimal set of mutations to arrive from one sequence to another).\n",
    "\n",
    "Dynamic programming process:\n",
    "![needleman_wunsch_intuition.png](needleman_wunsch_intuition.png)\n",
    "\n",
    "Article explaining this in depth: https://studentsxstudents.com/coding-global-sequence-alignment-using-the-needleman-wunsch-algorithm-d47971ebbe5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "36cf8d36",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nw(x, y, match = 1, mismatch = 1, gap = 2):\n",
    "    nx = len(x) # length of first base sequence\n",
    "    ny = len(y) # length of second base sequence\n",
    "    \n",
    "    # Initialization process - forming the base matrix\n",
    "    F = np.zeros((nx + 1, ny + 1)) # an array \n",
    "    F[:,0] = np.linspace(0, -gap*nx, nx + 1)\n",
    "    F[0,:] = np.linspace(0, -gap*ny, ny + 1)\n",
    "    \n",
    "    # Pointers to trace through an optimal aligment.\n",
    "    P = np.zeros((nx + 1, ny + 1))\n",
    "    P[:,0] = 3\n",
    "    P[0,:] = 4\n",
    "    \n",
    "    print(F)\n",
    "    print(P)\n",
    "    # Temporary scores for aligning base x and base y.\n",
    "    t = np.zeros(3)\n",
    "    for i in range(nx):\n",
    "        for j in range(ny):\n",
    "            \n",
    "            # Iteration step: take the max (inserting gap in first sequence, inserting gap in second sequence, match or mutation)\n",
    "            if x[i] == y[j]:\n",
    "                t[0] = F[i,j] + match\n",
    "            else:\n",
    "                t[0] = F[i,j] - mismatch\n",
    "            \n",
    "            # Inserting gap in first sequence\n",
    "            t[1] = F[i,j+1] - gap\n",
    "            # Inserting gap in second sequence\n",
    "            t[2] = F[i+1,j] - gap\n",
    "            tmax = np.max(t)\n",
    "            \n",
    "            F[i+1,j+1] = tmax\n",
    "            if t[0] == tmax:\n",
    "                P[i+1,j+1] += 2\n",
    "                \n",
    "            # Higher weights for inserting gaps rather than matches/mismatches\n",
    "            if t[1] == tmax:\n",
    "                P[i+1,j+1] += 3\n",
    "            if t[2] == tmax:\n",
    "                P[i+1,j+1] += 4\n",
    "\n",
    "    # Trace through an optimal alignment.\n",
    "    i = nx\n",
    "    j = ny\n",
    "    rx = []\n",
    "    ry = []\n",
    "    \n",
    "    tracer_matrix = np.zeros((nx+1, ny+1))\n",
    "    while i > 0 or j > 0:\n",
    "        tracer_matrix[i, j] = -1\n",
    "        \n",
    "        if P[i,j] in [2, 5, 6, 9]:\n",
    "            rx.append(x[i-1])\n",
    "            ry.append(y[j-1])\n",
    "            \n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        \n",
    "        # if there's a gap in the first sequence\n",
    "        elif P[i,j] in [3, 7]:\n",
    "            rx.append(x[i-1])\n",
    "            ry.append('-')\n",
    "            i -= 1\n",
    "            \n",
    "        # if there's a gap in the second sequence\n",
    "        elif P[i,j] in [4]:\n",
    "            rx.append('-')\n",
    "            ry.append(y[j-1])\n",
    "            j -= 1\n",
    "    \n",
    "    \n",
    "    print(\"Pointer matrix:\")\n",
    "    pointer_matrix = add_bases(x, y, P)\n",
    "    print_matrix(pointer_matrix)\n",
    "    \n",
    "    print()\n",
    "    print(\"Tracer matrix:\")\n",
    "    tracer_matrix = add_bases(x, y, tracer_matrix)\n",
    "    print_matrix(tracer_matrix)\n",
    "    \n",
    "    # Reverse the strings.\n",
    "    print()\n",
    "    print(\"Final result:\")\n",
    "    \n",
    "    rx = ''.join(rx)[::-1]\n",
    "    ry = ''.join(ry)[::-1]\n",
    "    \n",
    "    px = \"Sequence 1: \" + rx\n",
    "    py = \"Sequence 2: \" + ry\n",
    "    return ['\\n'.join([px, py]), rx, ry]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55b28d81",
   "metadata": {},
   "source": [
    "# Random Sequences\n",
    "\n",
    "This section of the code randomly chooses As, Ts, Cs, and Gs in order to develop a comprehensive understanding of how randomness plays a role in the N-W algorithm. As evidenced by the code below, it seems that randomness would simply confuse the N-W algorithm since there is no optimal mutation path to get from one sequence to another."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "1015a9ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 0. -2. -4. -6. -8.]\n",
      " [-2.  0.  0.  0.  0.]\n",
      " [-4.  0.  0.  0.  0.]\n",
      " [-6.  0.  0.  0.  0.]\n",
      " [-8.  0.  0.  0.  0.]]\n",
      "[[4. 4. 4. 4. 4.]\n",
      " [3. 0. 0. 0. 0.]\n",
      " [3. 0. 0. 0. 0.]\n",
      " [3. 0. 0. 0. 0.]\n",
      " [3. 0. 0. 0. 0.]]\n",
      "Pointer matrix:\n",
      " \t   \tT  \tG  \tC  \tC  \n",
      " \t4.0\t4.0\t4.0\t4.0\t4.0\n",
      "C\t3.0\t2.0\t6.0\t2.0\t6.0\n",
      "C\t3.0\t5.0\t2.0\t2.0\t2.0\n",
      "G\t3.0\t5.0\t2.0\t2.0\t2.0\n",
      "T\t3.0\t2.0\t3.0\t2.0\t2.0\n",
      "\n",
      "Tracer matrix:\n",
      " \t   \tT   \tG   \tC   \tC   \n",
      " \t0.0\t0.0 \t0.0 \t0.0 \t0.0 \n",
      "C\t0.0\t-1.0\t0.0 \t0.0 \t0.0 \n",
      "C\t0.0\t0.0 \t-1.0\t0.0 \t0.0 \n",
      "G\t0.0\t0.0 \t0.0 \t-1.0\t0.0 \n",
      "T\t0.0\t0.0 \t0.0 \t0.0 \t-1.0\n",
      "\n",
      "Final result:\n",
      "Sequence 1: CCGT\n",
      "Sequence 2: TGCC\n"
     ]
    }
   ],
   "source": [
    "np.random.seed(12)\n",
    "\n",
    "# For random sequences (sequence alignment not optimal)\n",
    "x = np.random.choice(['A', 'T', 'G', 'C'], 4)\n",
    "y = np.random.choice(['A', 'T', 'G', 'C'], 4)\n",
    "\n",
    "printseq, seq1, seq2 = nw(x,y)\n",
    "print(printseq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c575564",
   "metadata": {},
   "source": [
    "# With E.Coli Data\n",
    "\n",
    "There is an evident use case in utilizing the Needleman-Wunsch algorithm in finding differences/similarities in parts of the E. coli genetic sequence.\n",
    "\n",
    "Here, we can discover the exact set of mutations, additions, and deletions that allowed us to arrive from one sequence to another. If you’re interested in accessing the data, the text file “e_coli_seq.txt” should provide the necessary genetic info to test out different E. coli sequences.\n",
    "\n",
    "Below is the code testing out different E. coli sequences with one another to check for mutations and alterations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "6e64b67b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dna = open(\"e_coli_seq.txt\")\n",
    "query = open(\"query.txt\")\n",
    "\n",
    "seq_arrange = []\n",
    "for line in dna:\n",
    "    if line[0] != '>':\n",
    "        end = len(line)\n",
    "        seq_arrange.append(line[:-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "18b6081a",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_arrange = [i[:13] for i in seq_arrange]\n",
    "\n",
    "first_index = 0\n",
    "second_index = 0\n",
    "\n",
    "while (first_index == second_index):\n",
    "    first_index = randint(0, len(seq_arrange))\n",
    "    second_index = randint(0, len(seq_arrange))\n",
    "first_seq = seq_arrange[first_index]\n",
    "second_seq = seq_arrange[second_index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "8fc2bb07",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pointer matrix:\n",
      " \t   \tC  \tT  \tC  \tT  \tC  \tA  \tT  \tG  \tG  \tA  \tA  \tG  \tT  \n",
      " \t4.0\t4.0\t4.0\t4.0\t4.0\t4.0\t4.0\t4.0\t4.0\t4.0\t4.0\t4.0\t4.0\t4.0\n",
      "C\t3.0\t2.0\t4.0\t6.0\t4.0\t6.0\t4.0\t4.0\t4.0\t4.0\t4.0\t4.0\t4.0\t4.0\n",
      "C\t3.0\t5.0\t2.0\t2.0\t4.0\t6.0\t4.0\t4.0\t4.0\t4.0\t4.0\t4.0\t4.0\t4.0\n",
      "G\t3.0\t3.0\t5.0\t2.0\t2.0\t6.0\t6.0\t6.0\t2.0\t6.0\t4.0\t4.0\t6.0\t4.0\n",
      "G\t3.0\t3.0\t5.0\t5.0\t2.0\t2.0\t6.0\t6.0\t2.0\t2.0\t4.0\t4.0\t6.0\t4.0\n",
      "T\t3.0\t3.0\t2.0\t5.0\t2.0\t2.0\t2.0\t2.0\t4.0\t6.0\t2.0\t6.0\t6.0\t2.0\n",
      "T\t3.0\t3.0\t5.0\t2.0\t5.0\t2.0\t2.0\t2.0\t6.0\t6.0\t6.0\t2.0\t6.0\t2.0\n",
      "A\t3.0\t3.0\t3.0\t5.0\t5.0\t5.0\t2.0\t7.0\t2.0\t6.0\t2.0\t6.0\t6.0\t6.0\n",
      "T\t3.0\t3.0\t5.0\t5.0\t2.0\t5.0\t3.0\t2.0\t4.0\t2.0\t6.0\t2.0\t6.0\t2.0\n",
      "T\t3.0\t3.0\t5.0\t5.0\t5.0\t2.0\t3.0\t5.0\t2.0\t6.0\t2.0\t6.0\t2.0\t2.0\n",
      "G\t3.0\t3.0\t3.0\t5.0\t3.0\t5.0\t5.0\t3.0\t2.0\t2.0\t4.0\t4.0\t2.0\t6.0\n",
      "A\t3.0\t3.0\t3.0\t5.0\t3.0\t5.0\t2.0\t3.0\t3.0\t5.0\t2.0\t6.0\t4.0\t4.0\n",
      "G\t3.0\t3.0\t3.0\t5.0\t3.0\t5.0\t3.0\t5.0\t5.0\t2.0\t3.0\t2.0\t2.0\t4.0\n",
      "A\t3.0\t3.0\t3.0\t5.0\t3.0\t5.0\t5.0\t5.0\t3.0\t3.0\t2.0\t2.0\t2.0\t2.0\n",
      "\n",
      "Tracer matrix:\n",
      " \t    \tC   \tT   \tC   \tT   \tC   \tA   \tT   \tG   \tG   \tA   \tA   \tG   \tT   \n",
      " \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \n",
      "C\t-1.0\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \n",
      "C\t0.0 \t-1.0\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \n",
      "G\t0.0 \t0.0 \t-1.0\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \n",
      "G\t0.0 \t0.0 \t0.0 \t-1.0\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \n",
      "T\t0.0 \t0.0 \t0.0 \t0.0 \t-1.0\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \n",
      "T\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t-1.0\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \n",
      "A\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t-1.0\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \n",
      "T\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t-1.0\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \n",
      "T\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t-1.0\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \n",
      "G\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t-1.0\t-1.0\t0.0 \t0.0 \t0.0 \n",
      "A\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t-1.0\t0.0 \t0.0 \n",
      "G\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t-1.0\t0.0 \n",
      "A\t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t0.0 \t-1.0\n",
      "\n",
      "Final result:\n",
      "Sequence 1: CCGGTTATTG-AGA\n",
      "Sequence 2: -CTCTCATGGAAGT\n"
     ]
    }
   ],
   "source": [
    "printseq, seq1, seq2 = nw(first_seq, second_seq)\n",
    "print(printseq)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
